File: layoutz-hs/examples/InlineLoadingDemo.hs
{-# LANGUAGE OverloadedStrings #-}

{-|
Inline loading demo - progress bars that render in-place without
clearing the screen. The user prompt and prior output stay visible.

Run with: cabal run inline-loading-demo
-}

module Main where

import Layoutz

data S = S { prog :: Double, done :: Int }
data M = Tick

progressBar :: String -> Double -> LayoutzApp S M
progressBar label speed = LayoutzApp
  { appInit = (S 0.0 0, CmdNone)
  , appUpdate = \Tick st ->
      if done st > 20     then (st, CmdExit)
      else if prog st >= 1.0 then (st { done = done st + 1 }, CmdNone)
      else (st { prog = min 1.0 (prog st + speed) }, CmdNone)
  , appSubscriptions = \_ -> subEveryMs 16 Tick
  , appView = \st ->
      let w = 40
          filled = floor (prog st * fromIntegral w) :: Int
          bar = map (\i ->
            let r' = fromIntegral i / fromIntegral w :: Double
                r = floor (r' * 180) + 50
                g = floor ((1 - r') * 200) + 55
            in withColor (ColorTrue r g 255) $ text "█"
            ) [0 .. filled - 1]
          empty = replicate (w - filled) (withColor ColorBrightBlack $ text "░")
          pct = show (floor (prog st * 100) :: Int) ++ "%"
      in layout [ tightRow (bar ++ empty)
                , withColor ColorBrightCyan $ text $ label <> " " <> pct
                ]
  }

main :: IO ()
main = do
  putStrLn "Just a normal process"
  runInline (progressBar "Fetching deps..." 0.018)
  runInline (progressBar "Building..."      0.010)
  runInline (progressBar "Linking..."       0.025)

File: layoutz-ocaml/inline_loading_demo.ml
open Layoutz

(* Inline loading demo — progress bars that render in-place without
   clearing the screen. Prior output stays visible.

   Run with: dune exec ./inline_loading_demo.exe *)

type load_state = { progress: float; done_ticks: int }
type load_msg = Tick

let inline_options =
  { default_options with clear_on_start = false; clear_on_exit = false }

let progress_bar label speed =
  {
    init = ({ progress = 0.0; done_ticks = 0 }, CmdNone);
    update =
      (fun Tick st ->
        if st.done_ticks > 20
        then (st, CmdExit)
        else if st.progress >= 1.0
        then ({ st with done_ticks = st.done_ticks + 1 }, CmdNone)
        else ({ st with progress = min 1.0 (st.progress +. speed) }, CmdNone));
    subscriptions = (fun _ -> sub_every_ms 16 Tick);
    view =
      (fun st ->
        let w = 40 in
        let filled = int_of_float (st.progress *. float_of_int w) in
        let bar =
          List.init w (fun i ->
              if i < filled
              then
                let ratio = float_of_int i /. float_of_int w in
                let r = int_of_float (ratio *. 180.0) + 50 in
                let g = int_of_float ((1.0 -. ratio) *. 200.0) + 55 in
                s "█" |> colorRGB r g 255
              else s "░" |> colorBrightBlack)
        in
        let pct = Printf.sprintf "%d%%" (int_of_float (st.progress *. 100.0)) in
        layout [ tightRow bar; s (label ^ " " ^ pct) |> colorBrightCyan ]);
  }

let sleep_ms ms = Unix.sleepf (float_of_int ms /. 1000.0)

let () =
  print_endline "hello from a normal process";
  sleep_ms 800;
  print_endline "doing some work...";
  sleep_ms 700;
  print_endline "now watch this:";
  sleep_ms 800;
  print_newline ();
  run_app ~options:inline_options (progress_bar "Fetching deps..." 0.018);
  run_app ~options:inline_options (progress_bar "Building..." 0.010);
  run_app ~options:inline_options (progress_bar "Linking..." 0.025);
  print_newline ();
  print_endline "back to normal output"
